api/hds.proto

```diff
@@ -0,0 +1,105 @@
+  //    a. Endpoints to health checks
+  //    b. Per cluster configuration change
+  // 3. Envoy creates a health probe based on the HealthCheck config and sends
+  //    it to endpoint(ip:port) of Host 1 and 2. Based on the HealthCheck
+  //    configuration Envoy waits upon the arrival of the probe response and
+  //    looks at the content of the response to decide whether the endpoint is
+  //    healthy or not. If a response hasn’t been received within the timeout
+  //    interval, the endpoint health status is considered TIMEOUT.
+  // 4. Envoy reports results back in an EndpointHealthResponse message.
+  //    Envoy streams responses as often as the interval configured by the
+  //    management server in HealthCheckSpecifier.
+  // 5. The management Server collects health statuses for all endpoints in the
+  //    cluster (for all clusters) and uses this information to construct
+  //    EndpointDiscoveryResponse messages.
+  // 6. Once Envoy has a list of upstream endpoints to send traffic to, it load
+  //    balances traffic to them without additional health checking. It may
+  //    use inline healthcheck (i.e. consider endpoint UNHEALTHY if connection
+  //    failed to a particular endpoint to account for health status propagation
+  //    delay between HDS and EDS).
+  // By default, can_healthcheck is true. If can_healthcheck is false, Cluster
+  // configuration may not contain HealthCheck message.
+  // TODO(htuch): How is can_healthcheck communicated to CDS to ensure the above
+  // invariant?
+  // TODO(htuch): Add @amb67's diagram.
+  rpc CheckHealth(stream HealthCheckRequestOrEndpointHealthResponse)
+      returns (stream HealthCheckSpecifier) {
+  }
+}
+
+// Defines supported protocols etc, so the management server can assign proper
+// endpoints to healthcheck.
+message Capability {
+  // TODO(htuch): Do we really need capability support? Won't all Envoys be
+  // capable of doing both HTTP and TCP?
+  enum Protocol {
+    HTTP = 0;
+    TCP = 1;
+  }
+  repeated Protocol health_check_protocol = 1;
+}
+
+message HealthCheckRequest {
+  Node node = 1;
+  Capability capability = 2;
+}
+
+message EndpointHealth {
+  Endpoint endpoint = 1;
+  HealthStatus health_status = 2;
+}
+
+message EndpointHealthResponse {
+  repeated EndpointHealth endpoints_health = 1;
+}
+
+message HealthCheckRequestOrEndpointHealthResponse {
+  // TODO(htuch): Slight change to Envoy -> management server message stream to
+  // make use of oneof for clarity, confirm with @amb67 this is reasonable.
+  oneof request_type {
+    HealthCheckRequest health_check_request = 1;
+    EndpointHealthResponse endpoint_health_response = 2;
+  }
+}
+
+message LocalityEndpoints {
+  Locality locality = 1;
+  repeated Endpoint endpoints = 2;
+}
+
+// TODO(htuch): Do we need the cluster name and Locality in HDS? Doesn't Envoy
+// already know about the cluster name and locality for each Endpoint from EDS?
```

Comment:
@amb67 Is this for local stats purpose only or is there some other reason?

On an unrelated topic, I'm starting to wonder about corner cases where the same Endpoint is placed in multiple localities (e.g. in zone A and subzone X and subzone Y) by EDS or HDS. Is this illegal?